### PR TITLE
[WIP] Implement mailchimp subscriptions

### PR DESCRIPTION
**Pull request by kytrinyx**
First a warning. If you google for the official ruby mailchimp client you get https://bitbucket.org/mailchimp/mailchimp-api-ruby

TL;DR: Don't use it.

Don't be fooled by the fact that the account name is mailchimp. MailChimp didn't even know this wrapper existed until I pinged one of their employees in a DM yesterday. The API that it talks to has been scrapped (https://devs.mailchimp.com/blog/sunsetting-previous-mailchimp-api-versions/).

If you want the even longer version of this story—the one with yaks—I'll tell you  sometime.

Anyway. There are no official API wrappers for MailChimp. We're only talking to one endpoint, so I just went with plain HTTP calls with Faraday. I stuck it all in its own class, I may have gone overboard with the methodization (random bits of data make me confused and nervous, so I err towards the verbose when talking to APIs that are unknown to me).

There are a few things I wasn't sure what to do with:

1. The default for the API (with an API key) is basic auth. The transport is over HTTPS, so that might not be as terrible as it sounds, but still. We may want to consider using oauth instead of the API key (docs here: http://developer.mailchimp.com/documentation/mailchimp/guides/how-to-use-oauth2/)
1. Error handling in the controller. It's kind of gross right now, but I really don't know what sort of things will typically go wrong. For now I'm just logging errors and apologizing and moving on.
1. I don't think we're showing flashes yet
1. I don't HTML and CSS very well
1. We need copy for the newsletter part

screen shot 2017-06-09 at 1 59 19 pm

Happy open source Friday! ✨ 💛  ✨

### CODE REVIEW

**Comment by kytrinyx** [comment]
I'm not sure why the build is failing—there's no error that I can see in the tests, it just says that the build script exited with 2.

**Comment by MikeMcQuaid** [comment]
> I'm not sure why the build is failing—there's no error that I can see in the tests, it just says that the build script exited with 2.

It's an issue with insufficient code coverage:
> Coverage (51.50%) is below the expected minimum coverage (80.00%).

File: README.md [hunk 1]
```diff
@@ -103,6 +103,17 @@ our browser to <http://localhost:3000>
 ./script/server
 ```
 
+### Newsletter
+
+If you are going to be developing or testing the MailChimp integration, you will need
+
+* a [MailChimp account](https://mailchimp.com)
+* an [API key](http://kb.mailchimp.com/integrations/api-integrations/about-api-keys)
+* a [list](http://kb.mailchimp.com/lists/manage-contacts/find-your-list-id)
```
Comment (MikeMcQuaid) [inline README.md]: `list id`?

File: app/controllers/subscriptions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
+require_relative "../models/mailchimp"
+
+class SubscriptionsController < ApplicationController
+  # Subscribe to the MailChimp newsletter.
+  def create
+    redirect_to request.referer if !current_user
+
+    error = false
+    begin
```
Comment (MikeMcQuaid) [inline app/controllers/subscriptions_controller.rb]: `error = begin` maybe and then use the same logic below minus the `error =`?

File: app/controllers/subscriptions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
+require_relative "../models/mailchimp"
+
+class SubscriptionsController < ApplicationController
+  # Subscribe to the MailChimp newsletter.
+  def create
+    redirect_to request.referer if !current_user
```
Comment (MikeMcQuaid) [inline app/controllers/subscriptions_controller.rb]: Check out `rubocop` locally; the CI didn't display it in the output because of the test coverage failure (if it'd be more helpful I can make rubocop either run first or ensure both always run?).

**Comment by MikeMcQuaid** [comment]
> Anyway. There are no official API wrappers for MailChimp. We're only talking to one endpoint, so I just went with plain HTTP calls with Faraday. I stuck it all in its own class, I may have gone overboard with the methodization (random bits of data make me confused and nervous, so I err towards the verbose when talking to APIs that are unknown to me).

:+1: :sob:

> The default for the API (with an API key) is basic auth. The transport is over HTTPS, so that might not be as terrible as it sounds, but still. 
> Error handling in the controller. It's kind of gross right now, but I really don't know what sort of things will typically go wrong. For now I'm just logging errors and apologizing and moving on.

These are fine with me as-is 👍 

> I don't think we're showing flashes yet

Yeh, would be good to get that before we 🚢 this. I can have a look if that'd be good.

> I don't HTML and CSS very well
> We need copy for the newsletter part

Similarly: I can take a look at these and push to your PR if that's 🆒 with you.

File: app/controllers/subscriptions_controller.rb (hunk 3 above)
Comment (kytrinyx) [inline app/controllers/subscriptions_controller.rb]: If CI can say what's wrong (whether that is test coverage or rubocop or failing tests or any of the above) that would be helpful.

File: app/controllers/subscriptions_controller.rb (hunk 2 above)
Comment (kytrinyx) [inline app/controllers/subscriptions_controller.rb]: ok, sure

File: README.md (hunk 1 above)
Comment (kytrinyx) [inline README.md]: yah

**Comment by kytrinyx** [comment]
For the test coverage, what do you think: VCR?

If you'd do the flashes and other HTML+CSS stuff that would be fantastic.

**Comment by MikeMcQuaid** [comment]
> For the test coverage, what do you think: VCR?

Yeh 👍 

> If you'd do the flashes and other HTML+CSS stuff that would be fantastic.

Sure!

File: app/controllers/subscriptions_controller.rb (hunk 3 above)
Comment (MikeMcQuaid) [inline app/controllers/subscriptions_controller.rb]: @kytrinyx Done in https://github.com/ossfriday/ossfriday/pull/68 so `script/cibuild` will always run both tests and rubocop even if the first fails.

Comment (kytrinyx) [inline app/controllers/subscriptions_controller.rb]: 🎉 awesome! Thank you

File: app/controllers/subscriptions_controller.rb (hunk 2 above)
Comment (kytrinyx) [inline app/controllers/subscriptions_controller.rb]: I tried something slightly different here. Would you take a look and see if you hate it? 🐩

**Comment by kytrinyx** [comment]
Rubocop is complaining about this:

```
app/controllers/subscriptions_controller.rb:6:21: C: Use && instead of and.
    redirect_to "/" and return unless current_user
                    ^^^
```

Rubocop is wrong. This is not a boolean `and`. It's control flow. Booleans should not be used for control flow.

**Comment by kytrinyx** [comment]
I added mailchimp env vars to the Travis repo settings, since the mailchimp test relies on them.

**Comment by kytrinyx** [comment]
They've classified boolean vs control flow as wonfix https://github.com/bbatsov/rubocop/issues/1288.

I've rewritten the code to be on multiple lines to avoid the whole thing :sob:

**Comment by MikeMcQuaid** [comment]
@kytrinyx Random question that just occurred to me: I was looking at the styling of the Open Source Guides mailing list subscribe and thought that they aren't doing anything server-side because it's hosted on Pages. Check out their form here: https://github.com/github/opensource.guide/blob/74ea45e7de269507cecf1f9989cb2c99f5ad36fa/_includes/footer.html#L22-L38

Would we be able to take a similar approach to avoid needing the server-side logic? Are there pros/cons of the server-side approach that I've missed? Thanks!

**Comment by MikeMcQuaid** [comment]
> Check out their form here: https://github.com/github/opensource.guide/blob/74ea45e7de269507cecf1f9989cb2c99f5ad36fa/_includes/footer.html#L22-L38

Less useful than that may be the PRs/commits with the diffs adding them:
https://github.com/github/opensource.guide/pull/110/files
https://github.com/github/open-source-survey/commit/ed999c4e254a7afccf29e8aba4f74ab64afe9143
https://github.com/github/open-source-survey/pull/87/files

Looks like we would need a form and JavaScript library.

**Comment by kytrinyx** [comment]
It didn't even occur to me that this was an option. That would be MUCH better.

In that case whoever has access to the mailing list needs to go generate the embed.
